ruby/ql/docs/flow_summaries.md

```diff
@@ -0,0 +1,205 @@
+# Flow summaries
+
+Flow summaries describe how data flows through methods whose definition is not
+included in the database. For example, methods in the standard library or a gem.
+
+Say we have the following code:
+
+```rb
+x = gets
+y = x.chomp
+system(y)
+```
+
+This code reads a line from STDIN, strips any trailing newlines, and executes it
+as a shell command. Assuming `x` is considered tainted, we want the argument `y`
+to be tainted in the call to `system`.
+
+`chomp` is a standard library method in the `String` class for which we
+have no source code, so we include a flow summary for it:
+
+```ql
+private class ChompSummary extends SimpleSummarizedCallable {
+  ChompSummary() { this = "chomp" }
+
+  override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+    input = "Argument[self]" and
+    output = "ReturnValue" and
+    preservesValue = false
+  }
+}
+```
+
+The shared dataflow library will use this summary to construct a fake definition
+for `chomp`. The behaviour of this definition depends on the body of
+`propagatesFlowExt`. In this case, the method will propagate taint flow from the
+`self` argument (i.e. the receiver) to the return value.
+
+If `preservesValue = true` then value flow is propagated. If it is `false` then
+only taint flow is propagated.
+
+Any call to `chomp` in the database will be translated, in the dataflow graph,
+to a call to this fake definition. 
+
+`input` and `output` define the "from" and "to" locations in the flow summary.
+They use a custom string-based syntax which is similar but not identical to
+Models as Data. These strings are often referred to as access paths.
```

Comment:
These strings should be identical to the `input` and `output` string in MaD.

Reply:
Ok, I see what you mean. However, I think of MaD collectively as all the components that make up a summary; but you are right that the access path component (which identifies the relevant calls) maps to API graphs, where as the input/output components have their own syntax.

Reply:
I don't think we should call the difference with the MaD `path` column just now (called the "identifying access path"). We can elaborate on this when documenting the MaD format.